### PR TITLE
Local jsx namespace

### PR DESCRIPTION
**Pull request by madou**
This pull request replaces the global JSX namespace augmentation with a local JSX namespace for the `css` prop. This is a breaking type change for TS users (if you aren't using TypeScript this change will not affect you) - however we would update our docs to use jsx pragma for consistency either way.

The benefits of this are:

- No longer polluting the global namespace
- Collisions in the global namespace are now impossible
- Having deeper more integrated types are now possible, for example it is now a type error when using CSS prop on a JSX element that doesn't take a className prop. We can extend this later on for tokens etc.

**Before**

```tsx
import '@compiled/react';

<div css={{ color: 'red' }} />

function Component(props: { children: string }) {
  return <div />
}

<Component css={{ color: 'red' }} /> // no error
```

**After (classic runtime)**

```tsx
/** @jsx jsx */
import { jsx } from '@compiled/react';

<div css={{ color: 'red' }} />

<Component css={{ color: 'red' }} /> 
           ^^^ error css prop doesn't exist
```

**After (automatic runtime)**

```tsx
/** @jsxImportSource @compiled/react */

<div css={{ color: 'red' }} />

<Component css={{ color: 'red' }} />
           ^^^ error css prop doesn't exist
```

**TODO**

- [x] Fix flow type errors
- [x] Add a changeset

**Follow up**

-  [x] Update lint rule to place JSX pragma (Done: https://github.com/atlassian-labs/compiled/pull/933 https://github.com/atlassian-labs/compiled/pull/932)
- [ ] Update codemod to place JSX pragma
- [ ] Fix type errors in examples (and see how we can typecheck them!)

### CODE REVIEW

File: .changeset/silver-dots-reply.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+---
+'@compiled/react': minor
+---
+
+**Breaking change:** When using the `css` prop with [TypeScript](https://www.typescriptlang.org) you now need to declare a jsx pragma enabling types for that module. Previously when importing the `@compiled/react` package the global JSX namespace would be polluted as a side effect potentially causing collisions with other CSS-in-JS libraries. Now thanks to the use of [locally scoped JSX namespaces](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html#locally-scoped-jsx-namespaces) the global JSX namespace will no longer be polluted.
```
Comment (DarkPurple141) [inline .changeset/silver-dots-reply.md:5]: jsx or JSX? Should we be consistent?

File: packages/babel-plugin-strip-runtime/src/__tests__/strip-runtime-source-code.test.ts [hunk 2]
```diff
@@ -96,8 +96,8 @@ describe('babel-plugin-strip-runtime using source code', () => {
         expect(actual).toMatchInlineSnapshot(`
           "/* app.tsx generated by @compiled/babel-plugin v0.0.0 */
 
-          import * as React from 'react';
           import { ax, ix } from '@compiled/react/runtime';
+          import * as React from 'react';
```
Comment (DarkPurple141) [inline packages/babel-plugin-strip-runtime/src/__tests__/strip-runtime-source-code.test.ts:100]: any reason these are no longer top imports?

File: packages/babel-plugin/src/types.tsx [hunk 3]
```diff
@@ -65,6 +65,14 @@ export interface State extends PluginPass {
     styled?: string;
   };
 
+  /**
+   * Details of pragmas that are currently enabled in the pass.
+   */
+  pragma: {
+    jsxImportSource?: boolean;
```
Comment (DarkPurple141) [inline packages/babel-plugin/src/types.tsx:72]: can both be enabled at the same time?

File: packages/babel-plugin/src/utils/append-runtime-imports.ts [hunk 4]
```diff
@@ -13,50 +13,55 @@ const importSpecifier = (name: string, localName?: string): t.ImportSpecifier =>
...
     });
   } else {
     // Add the runtime entrypoint module
-    path.insertBefore(
+    path.unshiftContainer(
+      'body',
       t.importDeclaration(
-        runtimeImportNames.map((runtimeImportName) => importSpecifier(runtimeImportName)),
-        t.stringLiteral(runtimeImportModuleName)
+        COMPILED_RUNTIME_IMPORTS.map((runtimeImportName) => importSpecifier(runtimeImportName)),
```
Comment (DarkPurple141) [inline packages/babel-plugin/src/utils/append-runtime-imports.ts:63]: nit: doesn't need an inline func, can just be `map(importSpecifier)`

**Review by DarkPurple141 — APPROVED** [review]

File: .changeset/silver-dots-reply.md (hunk 1 above)
Comment (madou) [reply .changeset/silver-dots-reply.md:5]: JSX

File: packages/babel-plugin-strip-runtime/src/__tests__/strip-runtime-source-code.test.ts (hunk 2 above)
Comment (madou) [reply packages/babel-plugin-strip-runtime/src/__tests__/strip-runtime-source-code.test.ts:100]: Oh, I can fix that. Sec.

File: packages/babel-plugin/src/types.tsx (hunk 3 above)
Comment (madou) [reply packages/babel-plugin/src/types.tsx:72]: Technically yes, but Babel or any bundler would throw

File: packages/babel-plugin/src/babel-plugin.tsx [hunk 5]
```diff
@@ -48,19 +51,45 @@ export default declare<State>((api) => {
...
+        enter(_, state) {
+          const { file } = state;
+
+          if (file.ast.comments) {
+            for (const comment of file.ast.comments) {
+              const jsxSourceMatches = JSX_SOURCE_ANNOTATION_REGEX.exec(comment.value);
+              const jsxMatches = JSX_ANNOTATION_REGEX.exec(comment.value);
+
+              if (jsxSourceMatches && jsxSourceMatches[1] === COMPILED_MODULE) {
+                // jsxImportSource pragma found - turn on CSS prop!
```
Comment (JakeLane) [inline packages/babel-plugin/src/babel-plugin.tsx:67]: This may resolve https://github.com/atlassian-labs/compiled/issues/836

**Review by JakeLane — APPROVED** [review]
awesome work!

File: packages/webpack-loader/src/__fixtures__/compiled-error.tsx [hunk 6]
```diff
@@ -1,5 +1,5 @@
-import { css } from '@compiled/react';
-import React from 'react';
+/** @jsx jsx */
+import { jsx, css } from '@compiled/react';
```
Comment (MonicaOlejniczak) [inline packages/webpack-loader/src/__fixtures__/compiled-error.tsx:2]: The order doesn't get autofixed? Bad rule

File: packages/webpack-loader/src/__fixtures__/local-styles.tsx [hunk 7]
```diff
@@ -1,13 +1,14 @@
...
+import { jsx, styled } from '@compiled/react';
+import { Fragment } from 'react';
 
 const Styled = styled.div({
   color: 'blue',
 });
 
 export const App = (): JSX.Element => (
-  <>
+  <Fragment>
```
Comment (MonicaOlejniczak) [inline packages/webpack-loader/src/__fixtures__/local-styles.tsx:10]: Does `<>` not work with the `jsx`?

File: tsconfig.paths.json [hunk 8]
```diff
@@ -7,6 +7,8 @@
       "@compiled/jest": ["packages/jest/src/index.tsx"],
       "@compiled/react": ["packages/react/src/index.tsx"],
       "@compiled/react/runtime": ["packages/react/src/runtime.tsx"],
+      "@compiled/react/jsx-runtime": ["packages/react/src/jsx/jsx-runtime.tsx"],
```
Comment (MonicaOlejniczak) [inline tsconfig.paths.json:10]: /nit these two should be above `react/runtime`

File: packages/babel-plugin/src/babel-plugin.tsx (hunk 5 above)
Comment (MonicaOlejniczak) [reply packages/babel-plugin/src/babel-plugin.tsx:67]: I don't think it will, but haven't dug deeply into it

File: packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  return transformSync(code, {
+    configFile: false,
+    babelrc: false,
+    plugins: [babelPlugin],
+  })?.code;
+};
+
+describe('local jsx namespace', () => {
+  it('should transform css prop using jsx import source', () => {
+    const result = transform(`
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx:15]: `result` to `actual` to match other test files

File: packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+      <CC>
+          <CS>{[_]}</CS>
+          {<div className={ax([\\"_syaz5scu\\"])}>hello</div>}
+        </CC>;"
+    `);
+  });
+
+  it('should transform css prop using jsx pragma', () => {
+    const result = transform(`
+      /** @jsxRuntime classic */
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx:37]: What happens with edge cases, should you test those too? i.e.`jsxRuntime automatic` with `@jsx jsx` and `@jsxImportSource` with the runtime specified, etc

File: stories/jsx-pragma.tsx [hunk 11]
```diff
@@ -0,0 +1,40 @@
+/** @jsxRuntime classic */
+/** @jsx jsx */
+// TypeScript thinks we're using the automatic runtime however we've forced Babel
```
Comment (MonicaOlejniczak) [inline stories/jsx-pragma.tsx:3]: How does babel handle `@jsxRuntime` exactly? I don't see any code references to it, and am unsure of where the magic comes from

File: stories/jsx-import-source-pragma.tsx [hunk 12]
```diff
@@ -0,0 +1,35 @@
+/** @jsxImportSource @compiled/react */
+
+export default {
```
Comment (MonicaOlejniczak) [inline stories/jsx-import-source-pragma.tsx:3]: Should these be stories or type tests? There are type tests for flow, you could use a similar format for typescript.

File: packages/react/src/__tests__/jest-matcher.test.tsx [hunk 13]
```diff
@@ -1,6 +1,7 @@
-import { styled } from '@compiled/react';
+/** @jsxRuntime classic */
+/** @jsx jsx */
+import { jsx, styled } from '@compiled/react';
```
Comment (MonicaOlejniczak) [inline packages/react/src/__tests__/jest-matcher.test.tsx:3]: Is there a way of enabling local namespaces without importing `jsx` and just using the pragma?

**Review by at-nathan — APPROVED** [review]

File: packages/react/src/__tests__/jest-matcher.test.tsx [hunk 14]
```diff
@@ -1,6 +1,7 @@
-import { styled } from '@compiled/react';
+/** @jsxRuntime classic */
```
[the remark this answers is not in the document]
Comment (madou) [reply packages/react/src/__tests__/jest-matcher.test.tsx:1]: Because the runtime is actually automatic, this forces Babel to use the classic runtime :-)

File: packages/webpack-loader/src/__fixtures__/local-styles.tsx (hunk 7 above)
Comment (madou) [reply packages/webpack-loader/src/__fixtures__/local-styles.tsx:10]: Yeah, exactly. Unless you specify the fragment pragma.

File: packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx (hunk 10 above)
Comment (madou) [reply packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx:37]: Build tools (Babel) throw - you can't define both.

File: stories/jsx-pragma.tsx (hunk 11 above)
Comment (madou) [reply stories/jsx-pragma.tsx:3]: It forces the runtime to be whatever is declared (classic or runtime). It's a babel thing unfortunately not supported by TS. Useful for testing :)

File: stories/jsx-import-source-pragma.tsx (hunk 12 above)
Comment (madou) [reply stories/jsx-import-source-pragma.tsx:3]: Both - there are tests testing these flows. They are also stories so it makes sure the runtime is actually set (because it imports from `@compiled/react/jsx-runtime`).

File: packages/react/src/__tests__/jest-matcher.test.tsx (hunk 13 above)
Comment (madou) [reply packages/react/src/__tests__/jest-matcher.test.tsx:3]: Using `jsxImportSource` and the automatic runtime 👍

File: packages/babel-plugin/src/utils/append-runtime-imports.ts [hunk 15]
```diff
@@ -13,50 +13,55 @@ const importSpecifier = (name: string, localName?: string): t.ImportSpecifier =>
...
-    .getAllPrevSiblings()
-    .concat(path.getAllNextSiblings())
-    .find(
-      (path) =>
-        t.isImportDeclaration(path.node) && path.node.source.value === runtimeImportModuleName
-    ) as NodePath<t.ImportDeclaration> | undefined;
+  const previouslyDeclaredRuntimeDeclaration = path
+    .get('body')
+    .find((childPath): childPath is NodePath<t.ImportDeclaration> => {
+      return (
```
[the remark this answers is not in the document]
Comment (madou) [reply packages/babel-plugin/src/utils/append-runtime-imports.ts:31]: I don't think there is a set standard, opportunity for a lint rule 😄

File: packages/react/src/__tests__/jest-matcher.test.tsx (hunk 13 above)
Comment (MonicaOlejniczak) [reply packages/react/src/__tests__/jest-matcher.test.tsx:3]: Yeah, so can we do that for this test?

File: stories/jsx-pragma.tsx (hunk 11 above)
Comment (MonicaOlejniczak) [reply stories/jsx-pragma.tsx:3]: Cool, found it here: https://babeljs.io/docs/en/babel-plugin-transform-react-jsx

File: packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx (hunk 10 above)
Comment (MonicaOlejniczak) [reply packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx:37]: So would you say that it's not worth having tests for these cases, because it's testing the external babel behaviour?

File: stories/jsx-pragma.tsx (hunk 11 above)
Comment (madou) [reply stories/jsx-pragma.tsx:3]: Oh! Sorry should have gotten it for you.

**Review by JakeLane — DISMISSED** [review]

File: packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx (hunk 10 above)
Comment (madou) [reply packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx:37]: It will throw when defining both - but still good to know what happens in a test (and if its behaviour changes in the future) - I'll add one an expect an exception.

File: packages/react/src/__tests__/jest-matcher.test.tsx (hunk 13 above)
Comment (madou) [reply packages/react/src/__tests__/jest-matcher.test.tsx:3]: Ah nice, sorted

File: packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx (hunk 10 above)
Comment (madou) [reply packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx:37]: Ah, it will throw when transforming the JSX elements with that plugin, however it won't throw in ours. I don't believe it's worth it to double up, what do you think?

Comment (MonicaOlejniczak) [reply packages/babel-plugin/src/css-prop/__tests__/jsx-pragma.test.tsx:37]: Yeah, all good to leave it as is

**Review by MonicaOlejniczak — APPROVED** [review]
